### PR TITLE
🐛 track MS with old labels after updating MD labels

### LINKED ISSUE #1341: MachineDeployment loses track of MachineSet when labels change
/kind bug

**What steps did you take and what happened:**

Given a MachineDeployment:

```
apiVersion: cluster.x-k8s.io/v1alpha2
kind: MachineDeployment
metadata:
  annotations:
    machinedeployment.clusters.k8s.io/revision: "1"
  name: nodes-0
spec:
  replicas: 1
  selector:
    matchLabels:
      set: node
  template:
    metadata:
      labels:
        cluster.x-k8s.io/cluster-name: cluster
        set: node
    spec:
      ...
```

I added (via `kubectl edit`) a label to both the `selector` and the `template`:

```
apiVersion: cluster.x-k8s.io/v1alpha2
kind: MachineDeployment
metadata:
  annotations:
    machinedeployment.clusters.k8s.io/revision: "1"
  name: nodes-0
spec:
  replicas: 1
  selector:
    matchLabels:
      set: node
      someLabel: foo
  template:
    metadata:
      labels:
        cluster.x-k8s.io/cluster-name: cluster
        set: node
        someLabel: foo
    spec:
      ...
```

The MachineDeployment controller created a new MachineSet matching the new template and scaled it up to the appropriate number of replicas, but never scaled down the old MachineSet.

**What did you expect to happen:**

_waves hands_ the deploy to have completed successfully, respecting whatever rolling update strategy is configured (I was testing with a single replica, but I would be surprised if it followed the strategy).

I *think* the way the Deployment controller works is to use the owner ref to infer that the old ReplicaSet is still part of the same "deployment": https://github.com/kubernetes/kubernetes/blob/06dc8cf4cb0d3bdc7b718392a66ca82678d9982f/pkg/controller/deployment/deployment_controller.go#L498-L499

**Anything else you would like to add:**

I haven't done a thorough look yet, but it seems like we might be able to add a "is controlled by this machinedeployment -> append to filtered" block before we check the label selector here: https://github.com/kubernetes-sigs/cluster-api/blob/e8858ef3fe0f66efa58f467f78c09ac698131797/controllers/machinedeployment_controller.go#L196 ?

**Environment:**

- Cluster-api version: recent v1alpha2 (and v1alpha1)

### PR DESCRIPTION
**Pull request by tahsinrahman**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1341

### CODE REVIEW

**Comment by tahsinrahman** [comment]
cc @sethp-nr

File: controllers/machinedeployment_controller.go [hunk 1]
```diff
@@ -193,7 +193,7 @@ func (r *MachineDeploymentReconciler) getMachineSetsForDeployment(d *clusterv1.M
 			continue
 		}
 
-		if !selector.Matches(labels.Set(ms.Labels)) {
+		if !selector.Matches(labels.Set(ms.Labels)) && !metav1.IsControlledBy(ms, d) {
```
Comment (ncdc) [inline controllers/machinedeployment_controller.go]: We already have the check for `!metav1.IsControlledBy(ms, d)` on line 211 - how is this different?

Comment (sethp-nr) [reply controllers/machinedeployment_controller.go]: I think this is what I was asking for in #1341. It's a double negative, but inverted it's "skip this machine set unless either the selector matches or it has a controller ref pointing to this machine deployment".

Comment (tahsinrahman) [reply controllers/machinedeployment_controller.go]: it doesn't ignore machinesets that do not have same labels after updating machinedeployment labels

Comment (ncdc) [reply controllers/machinedeployment_controller.go]: Thanks, I see the subtle difference now. Perhaps some comments would be in order?

Comment (tahsinrahman) [reply controllers/machinedeployment_controller.go]: I've added a comment

File: controllers/machinedeployment_controller_test.go [hunk 2]
```diff
@@ -206,6 +206,61 @@ var _ = Describe("MachineDeployment Reconciler", func() {
...
+		Eventually(func() bool {
+			// Set the all non-deleted machines as ready with a NodeRef, so the MachineSet controller can proceed
+			// to properly set AvailableReplicas.
+			machines := &clusterv1.MachineList{}
+			Expect(k8sClient.List(ctx, machines, client.InNamespace(namespace.Name))).NotTo(HaveOccurred())
+			for _, m := range machines.Items {
+				if !m.DeletionTimestamp.IsZero() {
+					continue
+				}
+				fakeInfrastructureRefReady(m.Spec.InfrastructureRef, infraResource)
```
Comment (ncdc) [inline controllers/machinedeployment_controller_test.go]: You'll likely also want to skip over Machines that belong to the original MachineSet - see #1378

Comment (tahsinrahman) [reply controllers/machinedeployment_controller_test.go]: thanks! it is working

**Comment by vincepri** [comment]
Great work @tahsinrahman !

**Pull request merged**